File: src/lib/csv-storage.js
import { CSV_COLUMNS, createInitialState } from "./data-schema.js";

export function isFileSystemAccessSupported() {
  return Boolean(window.showOpenFilePicker && window.showSaveFilePicker && window.FileSystemFileHandle);
}

export async function openCsvFile() {
  if (!isFileSystemAccessSupported()) {
    throw new Error("File System Access API is not supported. Use Chrome or Edge.");
  }

  const [fileHandle] = await window.showOpenFilePicker({
    multiple: false,
    types: [{ description: "CSV files", accept: { "text/csv": [".csv"] } }]
  });
  const file = await fileHandle.getFile();
  const csvText = await file.text();
  return { fileHandle, csvText };
}

export async function saveCsvFile(fileHandle, csvText) {
  if (!fileHandle?.createWritable) {
    throw new Error("No writable CSV file is open. Use Save As first.");
  }

  const writer = await fileHandle.createWritable();
  await writer.write(csvText);
  await writer.close();
}

export async function saveCsvFileAs(csvText, suggestedName = "expense-data.csv") {
  if (!isFileSystemAccessSupported()) {
    throw new Error("File System Access API is not supported. Use Chrome or Edge.");
  }

  const fileHandle = await window.showSaveFilePicker({
    suggestedName,
    types: [{ description: "CSV files", accept: { "text/csv": [".csv"] } }]
  });
  await saveCsvFile(fileHandle, csvText);
  return fileHandle;
}

export function parseAppCsv(csvText) {
  const { rows } = parseGenericCsv(csvText);
  const state = createInitialState();
  state.categories = [];
  state.subcategories = [];
  state.accounts = [];
  state.transactions = [];
  state.automaticTransactions = [];
  state.monthlySetup = [];

  for (const row of rows) {
    if (row.record_type === "meta") {
      state.selectedMonth = row.month || state.selectedMonth;
      state.currency = row.currency || state.currency;
      state.lastAccessedAt = row.last_accessed_at || "";
      state.lastAutomationRunAt = row.last_automation_run_at || "";
    }

    if (row.record_type === "category") {
      state.categories.push({
        id: row.id,
        name: row.category,
        group: row.group,
        defaultBudget: numberFrom(row.default_budget),
        taxBusinessReady: row.tax_business_ready || "No",
        notes: row.notes || ""
      });
    }

    if (row.record_type === "subcategory") {
      state.subcategories.push({
        id: row.id,
        category: row.category,
        name: row.subcategory,
        envelopeGroup: row.envelope_group || "",
        envelopeStyle: row.envelope_style || "Variable",
        defaultMonthlyTarget: numberFrom(row.default_monthly_target),
        notes: row.notes || ""
      });
    }

    if (row.record_type === "account") {
      state.accounts.push({
        id: row.id,
        name: row.account,
        type: row.account_type || "Bank",
        openingBalance: numberFrom(row.opening_balance),
        notes: row.notes || ""
      });
    }

    if (row.record_type === "transaction") {
      state.transactions.push({
        id: row.id,
        date: row.date,
        type: row.type || "Expense",
        category: row.category,
        subcategory: row.subcategory,
        account: row.account,
        merchantPayee: row.merchant_payee,
        description: row.description,
        amount: numberFrom(row.amount),
        currency: row.currency || state.currency,
        essential: booleanFrom(row.essential),
        month: row.month,
        reimbursable: booleanFrom(row.reimbursable),
        notes: row.notes || "",
        sourceRuleId: row.source_rule_id || ""
      });
    }

    if (row.record_type === "automatic_transaction") {
      state.automaticTransactions.push({
        id: row.id,
        enabled: row.enabled === "" ? true : booleanFrom(row.enabled),
        startDate: row.start_date,
        endDate: row.end_date || "",
        frequency: row.frequency || "Monthly",
        customInterval: numberFrom(row.custom_interval) || 1,
        customUnit: row.custom_unit || "Months",
        type: row.type || "Expense",
        category: row.category,
        subcategory: row.subcategory,
        account: row.account,
        merchantPayee: row.merchant_payee,
        description: row.description,
        amount: numberFrom(row.amount),
        currency: row.currency || state.currency,
        essential: booleanFrom(row.essential),
        reimbursable: booleanFrom(row.reimbursable),
        notes: row.notes || ""
      });
    }

    if (row.record_type === "monthly_setup") {
      state.monthlySetup.push({
        id: row.id,
        month: row.month,
        category: row.category,
        subcategory: row.subcategory,
        monthlyTarget: numberFrom(row.monthly_target),
        startingBalance: numberFrom(row.starting_balance),
        rollover: booleanFrom(row.rollover)
      });
    }
  }

  if (!state.categories.length || !state.subcategories.length || !state.accounts.length) {
    throw new Error("This CSV is missing required app records.");
  }

  return state;
}

export function serializeAppCsv(state) {
  const rows = [
    {
      record_type: "meta",
      id: "app",
      last_accessed_at: state.lastAccessedAt || "",
      last_automation_run_at: state.lastAutomationRunAt || "",
      month: state.selectedMonth,
      currency: state.currency
    },
    ...state.categories.map((item) => ({
      record_type: "category",
      id: item.id,
      category: item.name,
      group: item.group,
      default_budget: item.defaultBudget,
      tax_business_ready: item.taxBusinessReady,
      notes: item.notes
    })),
    ...state.subcategories.map((item) => ({
      record_type: "subcategory",
      id: item.id,
      category: item.category,
      subcategory: item.name,
      envelope_group: item.envelopeGroup,
      envelope_style: item.envelopeStyle,
      default_monthly_target: item.defaultMonthlyTarget,
      notes: item.notes
    })),
    ...state.accounts.map((item) => ({
      record_type: "account",
      id: item.id,
      account: item.name,
      account_type: item.type,
      opening_balance: item.openingBalance,
      notes: item.notes
    })),
    ...state.monthlySetup.map((item) => ({
      record_type: "monthly_setup",
      id: item.id,
      month: item.month,
      category: item.category,
      subcategory: item.subcategory,
      monthly_target: item.monthlyTarget,
      starting_balance: item.startingBalance,
      rollover: item.rollover
    })),
    ...(state.automaticTransactions || []).map((item) => ({
      record_type: "automatic_transaction",
      id: item.id,
      enabled: item.enabled,
      start_date: item.startDate,
      end_date: item.endDate,
      frequency: item.frequency,
      custom_interval: item.customInterval,
      custom_unit: item.customUnit,
      type: item.type,
      category: item.category,
      subcategory: item.subcategory,
      account: item.account,
      merchant_payee: item.merchantPayee,
      description: item.description,
      amount: item.amount,
      currency: item.currency,
      essential: item.essential,
      reimbursable: item.reimbursable,
      notes: item.notes
    })),
    ...state.transactions.map((item) => ({
      record_type: "transaction",
      id: item.id,
      date: item.date,
      month: item.month,
      type: item.type,
      category: item.category,
      subcategory: item.subcategory,
      account: item.account,
      merchant_payee: item.merchantPayee,
      description: item.description,
      amount: item.amount,
      currency: item.currency,
      essential: item.essential,
      reimbursable: item.reimbursable,
      notes: item.notes,
      source_rule_id: item.sourceRuleId || ""
    }))
  ];

  return [CSV_COLUMNS.join(","), ...rows.map((row) => CSV_COLUMNS.map((column) => escapeCsv(row[column] ?? "")).join(","))].join("\r\n");
}

export function parseGenericCsv(csvText) {
  const records = parseCsvRecords(csvText);
  const headers = records.shift()?.map((header) => header.trim()) || [];
  const rows = records
    .filter((record) => record.some((cell) => String(cell).trim() !== ""))
    .map((record) => Object.fromEntries(headers.map((header, index) => [header, record[index] ?? ""])));

  return { headers, rows };
}

function parseCsvRecords(csvText) {
  const records = [];
  let row = [];
  let cell = "";
  let inQuotes = false;

  for (let index = 0; index < csvText.length; index += 1) {
    const char = csvText[index];
    const next = csvText[index + 1];

    if (inQuotes && char === '"' && next === '"') {
      cell += '"';
      index += 1;
    } else if (char === '"') {
      inQuotes = !inQuotes;
    } else if (!inQuotes && char === ",") {
      row.push(cell);
      cell = "";
    } else if (!inQuotes && (char === "\n" || char === "\r")) {
      if (char === "\r" && next === "\n") {
        index += 1;
      }
      row.push(cell);
      records.push(row);
      row = [];
      cell = "";
    } else {
      cell += char;
    }
  }

  if (cell || row.length) {
    row.push(cell);
    records.push(row);
  }

  return records;
}

function escapeCsv(value) {
  const text = String(value);
  return /[",\r\n]/.test(text) ? `"${text.replaceAll('"', '""')}"` : text;
}

function numberFrom(value) {
  const number = Number(value);
  return Number.isFinite(number) ? number : 0;
}

function booleanFrom(value) {
  return ["true", "yes", "1", "y"].includes(String(value).toLowerCase());
}

File: backend/app/csv_storage.py
"""App CSV parsing, serialization, load, save, import, and export.

Preserves the existing normalized CSV columns and record types from the JS
frontend. Uses Python's csv module rather than manual string parsing.
"""

from __future__ import annotations

import csv
import io
from pathlib import Path

from .config import APP_CSV_PATH
from .schema_defaults import CSV_COLUMNS, create_initial_state


# ---------------------------------------------------------------------------
# Generic CSV parsing (replaces parseGenericCsv)
# ---------------------------------------------------------------------------

def parse_generic_csv(csv_text: str) -> tuple[list[str], list[dict[str, str]]]:
    """Return (headers, rows) from a generic CSV string."""
    reader = csv.reader(io.StringIO(csv_text))
    records = [row for row in reader]
    if not records:
        return [], []
    headers = [h.strip() for h in records[0]]
    rows = [
        dict(zip(headers, record))
        for record in records[1:]
        if any(cell.strip() for cell in record)
    ]
    return headers, rows


# ---------------------------------------------------------------------------
# Helper coercers (mirror JS numberFrom / booleanFrom)
# ---------------------------------------------------------------------------

def _number(value: str) -> float:
    try:
        n = float(value)
        return n if n == n and n not in (float("inf"), float("-inf")) else 0.0
    except (ValueError, TypeError):
        return 0.0


def _boolean(value: str) -> bool:
    return str(value).strip().lower() in ("true", "yes", "1", "y")


# ---------------------------------------------------------------------------
# Parsing app CSV
# ---------------------------------------------------------------------------

def parse_app_csv(csv_text: str) -> dict:
    """Parse an app CSV string into a state dict matching ExpenseState shape."""
    _headers, rows = parse_generic_csv(csv_text)
    state = create_initial_state()
    state["categories"] = []
    state["subcategories"] = []
    state["accounts"] = []
    state["transactions"] = []
    state["automaticTransactions"] = []
    state["monthlySetup"] = []

    for row in rows:
        rt = row.get("record_type", "")

        if rt == "meta":
            state["selectedMonth"] = row.get("month") or state["selectedMonth"]
            state["currency"] = row.get("currency") or state["currency"]
            state["lastAccessedAt"] = row.get("last_accessed_at") or ""
            state["lastAutomationRunAt"] = row.get("last_automation_run_at") or ""

        elif rt == "category":
            state["categories"].append({
                "id": row.get("id", ""),
                "name": row.get("category", ""),
                "group": row.get("group", ""),
                "defaultBudget": _number(row.get("default_budget", "0")),
                "taxBusinessReady": row.get("tax_business_ready") or "No",
                "notes": row.get("notes") or "",
            })

        elif rt == "subcategory":
            state["subcategories"].append({
                "id": row.get("id", ""),
                "category": row.get("category", ""),
                "name": row.get("subcategory", ""),
                "envelopeGroup": row.get("envelope_group") or "",
                "envelopeStyle": row.get("envelope_style") or "Variable",
                "defaultMonthlyTarget": _number(row.get("default_monthly_target", "0")),
                "notes": row.get("notes") or "",
            })

        elif rt == "account":
            state["accounts"].append({
                "id": row.get("id", ""),
                "name": row.get("account", ""),
                "type": row.get("account_type") or "Bank",
                "openingBalance": _number(row.get("opening_balance", "0")),
                "notes": row.get("notes") or "",
            })

        elif rt == "transaction":
            state["transactions"].append({
                "id": row.get("id", ""),
                "date": row.get("date", ""),
                "type": row.get("type") or "Expense",
                "category": row.get("category", ""),
                "subcategory": row.get("subcategory", ""),
                "account": row.get("account", ""),
                "merchantPayee": row.get("merchant_payee") or "",
                "description": row.get("description", ""),
                "amount": _number(row.get("amount", "0")),
                "currency": row.get("currency") or state["currency"],
                "essential": _boolean(row.get("essential", "")),
                "month": row.get("month", ""),
                "reimbursable": _boolean(row.get("reimbursable", "")),
                "notes": row.get("notes") or "",
                "sourceRuleId": row.get("source_rule_id") or "",
            })

        elif rt == "automatic_transaction":
            state["automaticTransactions"].append({
                "id": row.get("id", ""),
                "enabled": True if row.get("enabled") == "" else _boolean(row.get("enabled", "true")),
                "startDate": row.get("start_date", ""),
                "endDate": row.get("end_date") or "",
                "frequency": row.get("frequency") or "Monthly",
                "customInterval": int(_number(row.get("custom_interval", "1"))) or 1,
                "customUnit": row.get("custom_unit") or "Months",
                "type": row.get("type") or "Expense",
                "category": row.get("category", ""),
                "subcategory": row.get("subcategory", ""),
                "account": row.get("account", ""),
                "merchantPayee": row.get("merchant_payee") or "",
                "description": row.get("description", ""),
                "amount": _number(row.get("amount", "0")),
                "currency": row.get("currency") or state["currency"],
                "essential": _boolean(row.get("essential", "")),
                "reimbursable": _boolean(row.get("reimbursable", "")),
                "notes": row.get("notes") or "",
            })

        elif rt == "monthly_setup":
            state["monthlySetup"].append({
                "id": row.get("id", ""),
                "month": row.get("month", ""),
                "category": row.get("category", ""),
                "subcategory": row.get("subcategory", ""),
                "monthlyTarget": _number(row.get("monthly_target", "0")),
                "startingBalance": _number(row.get("starting_balance", "0")),
                "rollover": _boolean(row.get("rollover", "")),
            })

    if not state["categories"] or not state["subcategories"] or not state["accounts"]:
        raise ValueError("This CSV is missing required app records.")

    return state


# ---------------------------------------------------------------------------
# Serializing app CSV
# ---------------------------------------------------------------------------

def _escape_csv(value: str) -> str:
    """CSV-escape a single field value (mirrors JS escapeCsv)."""
    text = str(value)
    if any(ch in text for ch in ('"', ",", "\r", "\n")):
        return '"' + text.replace('"', '""') + '"'
    return text


def _make_row(**values) -> list[str]:
    """Build a CSV row list with exactly len(CSV_COLUMNS) entries.

    All columns default to empty string; only keys in ``values`` are set.
    """
    row = [values.get(col, "") for col in CSV_COLUMNS]
    # Convert booleans to lowercase strings
    for i, val in enumerate(row):
        if isinstance(val, bool):
            row[i] = str(val).lower()
    return row


def serialize_app_csv(state: dict) -> str:
    """Serialize state dict into the normalized multi-record-type app CSV string.

    Every row is guaranteed to have exactly len(CSV_COLUMNS) fields.
    """
    currency = state.get("currency", "USD")
    rows: list[list[str]] = []

    # meta
    rows.append(_make_row(
        record_type="meta",
        id="app",
        last_accessed_at=str(state.get("lastAccessedAt", "")),
        last_automation_run_at=str(state.get("lastAutomationRunAt", "")),
        month=state.get("selectedMonth", ""),
        currency=currency,
    ))

    # categories
    for item in state.get("categories", []):
        rows.append(_make_row(
            record_type="category",
            id=item.get("id", ""),
            category=item.get("name", ""),
            group=item.get("group", ""),
            default_budget=str(item.get("defaultBudget", 0)),
            tax_business_ready=item.get("taxBusinessReady", "No"),
            notes=item.get("notes", ""),
        ))

    # subcategories
    for item in state.get("subcategories", []):
        rows.append(_make_row(
            record_type="subcategory",
            id=item.get("id", ""),
            category=item.get("category", ""),
            subcategory=item.get("name", ""),
            envelope_group=item.get("envelopeGroup", ""),
            envelope_style=item.get("envelopeStyle", "Variable"),
            default_monthly_target=str(item.get("defaultMonthlyTarget", 0)),
            notes=item.get("notes", ""),
        ))

    # accounts
    for item in state.get("accounts", []):
        rows.append(_make_row(
            record_type="account",
            id=item.get("id", ""),
            account=item.get("name", ""),
            account_type=item.get("type", "Bank"),
            opening_balance=str(item.get("openingBalance", 0)),
            notes=item.get("notes", ""),
        ))

    # monthly_setup
    for item in state.get("monthlySetup", []):
        rows.append(_make_row(
            record_type="monthly_setup",
            id=item.get("id", ""),
            month=item.get("month", ""),
            category=item.get("category", ""),
            subcategory=item.get("subcategory", ""),
            monthly_target=str(item.get("monthlyTarget", 0)),
            starting_balance=str(item.get("startingBalance", 0)),
            rollover=item.get("rollover", False),
        ))

    # automatic_transactions
    for item in state.get("automaticTransactions", []):
        rows.append(_make_row(
            record_type="automatic_transaction",
            id=item.get("id", ""),
            enabled=item.get("enabled", True),
            start_date=item.get("startDate", ""),
            end_date=item.get("endDate", ""),
            frequency=item.get("frequency", "Monthly"),
            custom_interval=str(item.get("customInterval", 1)),
            custom_unit=item.get("customUnit", "Months"),
            type=item.get("type", "Expense"),
            category=item.get("category", ""),
            subcategory=item.get("subcategory", ""),
            account=item.get("account", ""),
            merchant_payee=item.get("merchantPayee", ""),
            description=item.get("description", ""),
            amount=str(item.get("amount", 0)),
            currency=item.get("currency", currency),
            essential=item.get("essential", False),
            reimbursable=item.get("reimbursable", False),
            notes=item.get("notes", ""),
        ))

    # transactions
    for item in state.get("transactions", []):
        rows.append(_make_row(
            record_type="transaction",
            id=item.get("id", ""),
            date=item.get("date", ""),
            month=item.get("month", ""),
            type=item.get("type", "Expense"),
            category=item.get("category", ""),
            subcategory=item.get("subcategory", ""),
            account=item.get("account", ""),
            merchant_payee=item.get("merchantPayee", ""),
            description=item.get("description", ""),
            amount=str(item.get("amount", 0)),
            currency=item.get("currency", currency),
            essential=item.get("essential", False),
            reimbursable=item.get("reimbursable", False),
            notes=item.get("notes", ""),
            source_rule_id=item.get("sourceRuleId", ""),
        ))

    header = ",".join(CSV_COLUMNS)
    body_lines = [",".join(_escape_csv(cell) for cell in row) for row in rows]
    return header + "\r\n" + "\r\n".join(body_lines)


# ---------------------------------------------------------------------------
# File-system helpers
# ---------------------------------------------------------------------------

def load_state(path: Path | None = None) -> dict:
    """Read and parse the app CSV from disk."""
    target = path or APP_CSV_PATH
    if not target.exists():
        raise FileNotFoundError(f"Data file not found: {target}")
    csv_text = target.read_text(encoding="utf-8")
    return parse_app_csv(csv_text)


def save_state(state: dict, path: Path | None = None) -> None:
    """Serialize state and write to disk."""
    target = path or APP_CSV_PATH
    target.parent.mkdir(parents=True, exist_ok=True)
    csv_text = serialize_app_csv(state)
    target.write_text(csv_text, encoding="utf-8")


def ensure_data_file(path: Path | None = None) -> dict:
    """Return parsed state from disk or initialized defaults, creating if needed."""
    target = path or APP_CSV_PATH
    if not target.exists():
        state = create_initial_state()
        save_state(state, target)
        return state
    return load_state(target)
